lib/base_service.js

```diff
@@ -31,13 +32,46 @@ class BaseService {
         this._logger = null;
         this._metrics = null;
         this._ratelimiter = null;
-        this.serviceReturns = null;
+        this._serviceReturns = null;
+
+        // Avoid recursion if there are bugs in the logging code.
+        this._inLogger = false;
+
+        const logUnhandledException = (err) => {
+            if (!this._inLogger) {
+                this._inLogger = true;
+                if (!err.stack) {
+                    Error.captureStackTrace(err);
+                }
+                this._logger.log('fatal/service-runner/unhandled', err);
+                this._inLogger = false;
+            }
+        };
+
+        this._uncaughtExceptionHandler = (err) => {
+            logUnhandledException(err);
+            // Don't exit right away, allow logger to process message
+            setTimeout(() => {
+                process.exit(1);
+            }, 100);
+        };
+
+        this._unhandledRejectionHandler = (err) => {
+            if (!this._inLogger) {
+                this._inLogger = true;
+                this._logger.log('fatal/service-runner/unhandled', err);
+                this._inLogger = false;
+            }
```

Comment:
I believe that before this change both handlers were essentially calling the same `logUnhandledException` code, git the unhandled exception handler was exiting the process as well. Can you please elaborate on why you have changed that?

Also, if you do that you only need to declare 2 functions, not 3.

Reply:
I changed it to 2. That was by mistake.